Sources/Commands/SwiftRunTool.swift

```diff
@@ -0,0 +1,170 @@
+    var mode: RunMode {
+        // If we got version option, just print the version and exit.
+        if shouldPrintVersion {
+            return .version
+        }
+
+        return .run
+    }
+    
+    /// If the executable product should be built before running.
+    var shouldBuild = true
+
+    /// Build configuration.
+    var config: Build.Configuration = .debug
+    
+    /// The custom working directory the executable should be run in.
+    var exeWorkingDir: AbsolutePath?
+    
+    /// The executable product to run.
+    var executable: String?
+    
+    /// The arguments to pass to the executable.
+    var arguments: [String] = []
+}
+
+public enum RunMode {
+    case version
+    case run
+}
+
+/// swift-run tool namespace
+public class SwiftRunTool: SwiftTool<RunToolOptions> {
+
+   public convenience init(args: [String]) {
+       self.init(
+            toolName: "run",
+            usage: "[options]",
+            overview: "Build and run an executable product",
+            args: args
+        )
+    }
+
+    override func runImpl() throws {
+        switch options.mode {
+        case .version:
+            print(Versioning.currentVersion.completeDisplayString)
+
+        case .run:
+            let exePath = try buildIfNeeded()
+            try run(path: exePath)
+        }
+    }
+
+    /// Builds the "executable" target if enabled in options.
+    ///
+    /// - Returns: The path to the executable binary.
+    private func buildIfNeeded() throws -> AbsolutePath {
+        let graph = try loadPackageGraph()
+        if options.shouldBuild {
+            try build(graph: graph, includingTests: false, config: options.config)
+        }
+        
+        let executable = try findExecutable(in: graph)
+
+        return buildPath
+            .appending(RelativePath(options.config.dirname))
+            .appending(component: executable.name)
+    }
+    
+    private func findExecutable(in graph: PackageGraph) throws -> ResolvedProduct {
+        let executableProducts = graph.products.filter({ $0.type == .executable })
+        
+        guard executableProducts.count > 0 else {
+            throw RunError.noExecutableFound
+        }
+        
+        if let executable = options.executable {
+            guard let executableProduct = executableProducts.first(where: { $0.name == executable }) else {
+                throw RunError.executableNotFound(executable)
+            }
+            
+            return executableProduct
+        } else {
+            guard executableProducts.count == 1 else {
+                throw RunError.multipleExecutables
+            }
+            
+            return executableProducts.first!
+        }
+    }
+    
+    /// Executes and returns execution status. Prints output on standard streams.
+    private func run(path: AbsolutePath) throws {
+        let process = Process(arguments: [path.asString] + options.arguments, redirectOutput: false)
```

Comment:
On UNIX, we should `exec` the process rather than run it in a subprocess. These will ensure that we don't influence the programs behavior w.r.t. signal handling, buffering, etc.